stl/inc/algorithm

```diff
@@ -8004,6 +8541,323 @@ void stable_sort(_ExPo&& _Exec, _BidIt _First, _BidIt _Last) noexcept /* termina
+            }
+        }
+
+        template <class _It, class _Pr, class _Pj>
+        static void _Insertion_sort_isort_max_chunks(
+            _It _First, const _It _Last, iter_difference_t<_It> _Count, _Pr _Pred, _Pj _Proj) {
+            // insertion sort every chunk of distance _Isort_max<_It> in [_First, _Last)
+            _STL_INTERNAL_STATIC_ASSERT(random_access_iterator<_It>);
+            _STL_INTERNAL_STATIC_ASSERT(sortable<_It, _Pr, _Pj>);
+            _STL_INTERNAL_CHECK(_RANGES distance(_First, _Last) == _Count);
+
+            for (; _Isort_max<_It> < _Count; _Count -= _Isort_max<_It>) { // sort chunks
+                _First = _RANGES _Insertion_sort_common(_First, _First + _Isort_max<_It>, _Pred, _Proj);
+            }
+
+            // sort partial last chunk
+            _RANGES _Insertion_sort_common(_STD move(_First), _STD move(_Last), _STD move(_Pred), _STD move(_Proj));
+        }
+
+        template <class _It, class _Pr, class _Pj>
+        static void _Uninitialized_chunked_merge_common(_It _First, const _It _Last, iter_value_t<_It>* const _Dest,
+            iter_difference_t<_It> _Count, _Pr _Pred, _Pj _Proj) {
+            // move to uninitialized merging adjacent chunks of distance _Chunk
+            _STL_INTERNAL_STATIC_ASSERT(random_access_iterator<_It>);
+            _STL_INTERNAL_STATIC_ASSERT(sortable<_It, _Pr, _Pj>);
+            _STL_INTERNAL_CHECK(_RANGES distance(_First, _Last) == _Count);
+
+            _Uninitialized_backout<iter_value_t<_It>*> _Backout{_Dest};
+            const auto _Backout_end            = _Dest + _Count;
+            iter_difference_t<_It> _Chunk_size = _Isort_max<_It>;
+            while (_Chunk_size < _Count) {
+                _Count -= _Chunk_size;
+                const auto _Chunk2 = (_STD min)(_Chunk_size, _Count);
+                _Count -= _Chunk2;
+
+                const auto _Mid1  = _First + _Chunk_size;
+                const auto _Last1 = _Mid1 + _Chunk2;
+                const auto _Last2 = _Backout._Last + _Chunk_size + _Chunk2;
+                _Backout._Last = _Uninitialized_merge_move(_First, _Mid1, _Last1, _Backout._Last, _Last2, _Pred, _Proj);
+                _First         = _Last1;
+            }
+
+            // copy partial last chunk
+            _RANGES _Uninitialized_move_unchecked(_STD move(_First), _STD move(_Last), _Backout._Last, _Backout_end);
+            _Backout._Release();
+        }
+
+        template <class _It, class _Pr, class _Pj>
+        _NODISCARD static iter_value_t<_It>* _Uninitialized_merge_move(_It _First, _It _Mid, _It _Last,
+            iter_value_t<_It>* const _Dest, iter_value_t<_It>* const _Dest_last, _Pr _Pred, _Pj _Proj) {
+            // move merging ranges to uninitialized storage
+            _STL_INTERNAL_STATIC_ASSERT(random_access_iterator<_It>);
+            _STL_INTERNAL_STATIC_ASSERT(sortable<_It, _Pr, _Pj>);
+            _STL_INTERNAL_CHECK(_First != _Mid);
+            _STL_INTERNAL_CHECK(_Mid != _Last);

```

Comment:
I think we're missing a precondition here that `_Last - _First <= _Dest_last - _Dest`.

Reply:
Done